// plugins/plugin-details/index.js
import './plugin-details.css'

function getTemplateDetails({ data = [], placeholder = 'placeholder', selectedValue }) {

  const items = data.map(value => {
    let cls = ''
    if (value === selectedValue) {
      // placeholder = value
      cls = 'list-group-item-secondary'
    }
    return `<button type="button" class="list-group-item list-group-item-action py-2 border-0 border-bottom ${cls}" data-type="item" data-value="${value}">${value}</button>`
  })

  return `
    <div class="backdrop position-fixed top-0 start-0 end-0 bottom-0" data-type="backdrop"></div>
    <summary class="details-summary d-flex align-items-center justify-content-between h-100 px-3 rounded-2 border">
      <span class="me-2" data-type="value">${placeholder}</span>
      <i class="bi-caret-down-fill" data-type="arrow"></i>
    </summary>
    <div class="details-list list-group position-absolute top-100 end-0 shadow mt-2 border rounded-3">
      <div class="list-group-item py-2 border-0 border-bottom">выбрать ${placeholder}</div>
      ${items.join('')}
    </div>`
}

export default class Details {
  constructor(selector, options) {
    this.$root = selector // details
    this.options = options
    this.selectedValue = options.selectedValue

    this.render()
    this.setup()
  }

  render() {
    this.$root.classList.add('details', 'position-relative')
    this.$root.innerHTML = getTemplateDetails(this.options)
  }

  handlerClick = (event) => {
    this.$root.hasAttribute('open') ? this.close() : this.open()

    if (event.target.dataset.type === 'item') {
      event.preventDefault()
      let value = event.target.dataset.value
      this.select(value)
    } else if (event.target.dataset.type === 'backdrop') {
      this.close()
    }
  }

  get current() {
    return this.options.data.find(value => value === this.selectedValue)
  }

  select(value) {
    this.selectedValue = value
    this.$value.innerHTML = this.current

    this.$root.querySelectorAll('[data-type="item"]').forEach(value => value.classList.remove('list-group-item-secondary'))
    // this.$root.querySelector(`[data-value="${value}"]`).classList.add('list-group-item-secondary')
    this.options.onSelect ? this.options.onSelect(this.current) : null
    this.close()
  }

  setup() {
    this.$root.addEventListener('click', this.handlerClick)
    this.$arrow = this.$root.querySelector('[data-type="arrow"]')
    this.$value = this.$root.querySelector('[data-type="value"]')
  }

  destroy() {
    this.$root.removeEventListener('click', this.handlerClick)
    this.$root.remove()
  }

  open() {
    this.$arrow.classList.add('bi-caret-up-fill')
    this.$arrow.classList.remove('bi-caret-down-fill')
  }

  close() {
    this.$root.removeAttribute('open')
    this.$arrow.classList.remove('bi-caret-up-fill')
    this.$arrow.classList.add('bi-caret-down-fill')
  }
}
